### ISSUE TITLE
Sik-Ho Tsang | Review: Learning Word Embeddings Efficiently with Noise-Contrastive Estimation (NCE).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#62]
Sik-Ho Tsang. [Review: Learning Word Embeddings Efficiently with Noise-Contrastive Estimation (NCE)](https://sh-tsang.medium.com/review-learning-word-embeddings-efficiently-with-noise-contrastive-estimation-nce-dba1c345c153).

**Comment by NorbertZheng** [comment]
## Overview
Using Noise-Contrastive Estimation (NCE) for Efficient Learning.

In this story, Learning Word Embeddings Efficiently with Noise-Contrastive Estimation, (NCE), by DeepMind, is briefly reviewed.
 - Continuous-valued word embeddings learned by neural language models have recently been shown to capture semantic and syntactic information. However, large amount of data is needed to train which makes the training time long.

In this paper:
 - A simple and scalable Noise-Contrastive Estimation is used to **shorten the training time**.

This is a paper in 2013 NeurIPS with over 600 citations. Noise-Contrastive Estimation (NCE) is another basic concept of Contrastive Learning in self-supervised learning.

**Comment by NorbertZheng** [comment]
## Softmax in [Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa)
 - In [Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa), Softmax is used at the end for CBOW and Skip-gram models:

$$
P_{\theta}^{h}(w)=\frac{\exp(s_{\theta}(w,h))}{\sum_{w'}\exp(s_{\theta}(w',h))}.
$$

 - However, the denominator needs to sum up all exponential values for all words in the vocabulary
    - When there are millions of words, the process is slow.

**Comment by NorbertZheng** [comment]
## Noise-Contrastive Estimation (NCE)
### Definition of Noise-Contrastive Estimation (NCE)
NCE is based on
 - the reduction of density estimation to probabilistic binary classification.

The basic idea is to train a logistic regression classifier to discriminate between samples from the data distribution and samples from some “noise” distribution, based on the ratio of probabilities of the sample under the model and the noise distribution.
 - We would like to learn the distribution of words for some specific context $h$, denoted by $P_{h}(w)$, a binary classifier is learned to classify the positive samples, i.e. the training samples from that context distribution $P_{h}(w)$ and the negative examples, i.e. the samples from a noise distribution $P_{n}(w)$.
 - Assuming that noise samples are $k$ times more frequent than data samples, the probability that the given sample came from the data is (Correct labels $D=1$ ):

$$
P_{h;\theta}(D=1|w)=\frac{P_{h;\theta}(w)}{P_{h;\theta}((w)+kP_{n}(w)}=\sigma(\Delta s_{\theta}(w,h)),
$$

 - where $\sigma(x)$ is the logistic function and $\Delta s_{\theta}(w,h)=s_{\theta}(w,h)-log(kP_{n}(w))$ is the difference in the scores of word $$ under the model $\theta$ and the (scaled) noise distribution.
 - And $s_{\theta}(w,h)$ is the network model used in this paper.

By using NCE, the summation originated at the denominator of the Softmax function can be skipped. Because NCE is an **unnormalized model**.

NCE training time is linear in the number of noise samples and independent of the vocabulary size.

As we increase the number of noise samples $k$, this estimate approaches the likelihood gradient of the normalized model.

 - The model is optimized by maximizing the log-posterior probability of the correct labels $D$ averaged over the data and noise samples:

![image](https://user-images.githubusercontent.com/37235023/217812092-313f41d4-c804-4088-a0d8-c13679fdfa03.png)

 - where the expectation over the noise distribution is approximated by sampling. And the gradient is:

![image](https://user-images.githubusercontent.com/37235023/217812187-8f87a75d-0560-46a6-9846-2d95265d663c.png)

 - where the gradient involves a sum over $k$ noise samples instead of a sum over the entire vocabulary, making the NCE training time linear in the number of noise samples and independent of the vocabulary size.

**Comment by NorbertZheng** [comment]
### Log-Bilinear Language (LBL) Models
 - (The models here are not the main point to learn NCE.)
 - In this paper, the log-bilinear language model (LBL) which was proposed by Prof. Hinton in 2007, is used as baseline for development.
 - It is a very simple neural language model. The LBL model performs linear prediction in the word feature vector space and does not have non-linearities, instead of using any hidden layers.
 - Vector LBL (vLBL) is the model to learn words based on the surrounding words.
 - ivLBL is the inverse of LBL, which predicts words that are surrounding the current word. It is similar to Skip-gram model in [Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa) but without position-dependent weights.

**Comment by NorbertZheng** [comment]
## Experimental Results
 - Two analogy-based word similarity tasks released by Google and Microsoft Research (MSR) are tested. (Please feel free to read the datasets details and experimental setup in the paper.)

![image](https://user-images.githubusercontent.com/37235023/217812514-96064828-0e57-49b5-9f97-9a6069cfb76f.png)
Accuracy in percent on word similarity tasks.

 - NCEk denotes NCE training using k noise samples.
 - All NCE-trained models outperformed the Skip-gram in [Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa). Accuracy steadily increased with the number of noise samples used, as did the training time.
 - The best compromise between running time and performance seems to be achieved with 5 or 10 noise samples.

![image](https://user-images.githubusercontent.com/37235023/217812654-02c276bc-3bfe-4cbf-8d19-71b0983d5522.png)
Accuracy in percent on word similarity tasks for large models.

 - The 300D ivLBL model trained for just over a day, achieves accuracy scores 3–9 percentage points better than the 300D Skip-gram trained for almost twice as long.
 - The same model trained for four days achieves accuracy scores that are only 2–4 percentage points lower than those of the 1000D Skip-gram trained on four times as much data using 75 times as many CPU cycles.

![image](https://user-images.githubusercontent.com/37235023/217812751-12b1315f-f92e-4cae-baf2-98bc29f0393a.png)
Results for various models trained for 20 epochs on the 47M-word Gutenberg dataset using NCE5 with AdaGrad ((D) and (I) denote models with and without position-dependent weights respectively).

 - Surprisingly, the results show that representations learned with position-independent weights, designated with (I), tend to perform better than the ones learned with position-dependent weights.

![image](https://user-images.githubusercontent.com/37235023/217812908-bb42b5fb-8f21-4af1-a236-46fd0f162e1e.png)
Accuracy on the MSR Sentence Completion Challenge dataset.

 - Even the model with the lowest embedding dimensionality of 100, achieves 51.0% correct, compared to 48.0% correct reported in [Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa) for the Skip-gram model with 640D embeddings.
 - The 55.5% correct achieved by the model with 600D embeddings is also better than the best single-model score on this dataset in the literature (54.7% in [14]).

**Comment by NorbertZheng** [comment]
Another work, which is “Distributed Representations of Words and Phrases and their Compositionality” in 2013 NeurIPS which proposes the negative sampling. The idea is very similar to NCE, which will be talked about later.

**Comment by NorbertZheng** [comment]
## Reference
[2013 NeurIPS] [NCE]
[Learning Word Embeddings Efficiently with Noise-Contrastive Estimation](https://proceedings.neurips.cc/paper/2013/file/db2b4182156b2f1f817860ac9f409ad7-Paper.pdf)

**Comment by NorbertZheng** [comment]
- Language Model: 2007 [[Bengio TNN’07](https://sh-tsang.medium.com/review-adaptive-importance-sampling-to-accelerate-training-of-a-neural-probabilistic-language-e71244001b13)] 2013 [[Word2Vec](https://sh-tsang.medium.com/review-word2vec-efficient-estimation-of-word-representations-in-vector-space-f9dbe2145afa)] [[NCE](https://sh-tsang.medium.com/review-learning-word-embeddings-efficiently-with-noise-contrastive-estimation-nce-dba1c345c153)]
 - Machine Translation: 2014 [[Seq2Seq](https://sh-tsang.medium.com/review-seq2seq-sequence-to-sequence-learning-with-neural-networks-bcb84071a670)] [[RNN Encoder-Decoder](https://sh-tsang.medium.com/review-learning-phrase-representations-using-rnn-encoder-decoder-for-statistical-machine-d4f345cb4382)] 2015 [[Attention Decoder/RNNSearch](https://sh-tsang.medium.com/review-neural-machine-translation-by-jointly-learning-to-align-and-translate-3b381fc032e3)]
 - Image Captioning: 2015 [[m-RNN](https://sh-tsang.medium.com/review-deep-captioning-with-multimodal-recurrent-neural-networks-m-rnn-4417d82ef4f5)] [[R-CNN+BRNN](https://sh-tsang.medium.com/review-deep-visual-semantic-alignments-for-generating-image-descriptions-fbdd30428cb)] [[Show and Tell/NIC](https://sh-tsang.medium.com/2d3928a90306)]
